modules/sagemaker/sagemaker-templates-service-catalog/templates/model_deploy/seed_code/deploy_app/deploy_app/pipeline_stack.py

```diff
@@ -0,0 +1,195 @@
+                effect=iam.Effect.ALLOW,
+                resources=[
+                    # TODO: this is clearly over-permissive
+                    f"arn:{cdk.Aws.PARTITION}:iam::{constants.DEV_ACCOUNT_ID}:role/*",
+                    f"arn:{cdk.Aws.PARTITION}:iam::{constants.PRE_PROD_ACCOUNT_ID}:role/cdk*",
+                    f"arn:{cdk.Aws.PARTITION}:iam::{constants.PROD_ACCOUNT_ID}:role/cdk*",
+                ],
+            ),
+            iam.PolicyStatement(
+                actions=["ssm:GetParameter"],
+                resources=[
+                    f"arn:{cdk.Aws.PARTITION}:ssm:{constants.DEV_REGION}:{constants.DEV_ACCOUNT_ID}:parameter/*",
+                    f"arn:{cdk.Aws.PARTITION}:ssm:{constants.PRE_PROD_REGION}:{constants.PRE_PROD_ACCOUNT_ID}:parameter/*",
+                    f"arn:{cdk.Aws.PARTITION}:ssm:{constants.PROD_REGION}:{constants.PROD_ACCOUNT_ID}:parameter/*",
+                ],
+            ),
+        ],
+    )
+
+
+class PipelineStack(cdk.Stack):
+    def __init__(self, scope: Construct, construct_id: str, **kwargs: Any) -> None:
+        super().__init__(scope, construct_id, **kwargs)
+
+        repository = codecommit.Repository.from_repository_name(
+            self, "Repository", repository_name=f"{constants.PROJECT_NAME}-deploy"
+        )
+
+        codepipeline_role = iam.Role(
+            self,
+            "CodePipelineRole",
+            assumed_by=iam.ServicePrincipal("codepipeline.amazonaws.com"),
+            path="/service-role/",
+        )
+
+        codepipeline_role.attach_inline_policy(create_inline_policy(self, "DeployPipelinePolicy"))
+
+        synth_codebuild_role = iam.Role(
+            self,
+            "SynthStageRole",
+            assumed_by=iam.ServicePrincipal("codebuild.amazonaws.com"),
+            path="/service-role/",
+        )
+        synth_codebuild_role.attach_inline_policy(create_inline_policy(self, "SynthStagePolicy"))
+
+        pipeline = CodePipeline(
+            self,
+            "Pipeline",
+            pipeline_name=f"{constants.PROJECT_NAME}-pipeline",
+            synth=CodeBuildStep(
+                "Synth",
+                input=CodePipelineSource.code_commit(repository=repository, branch="main"),
+                install_commands=[
+                    "npm install -g aws-cdk",
+                ],
+                commands=["python -m pip install -r requirements.txt", 'cdk synth --app "python app.py" '],
+                role=synth_codebuild_role,
+                env=ENV,
+            ),
+            cross_account_keys=True,
+            self_mutation=True,
+            role=codepipeline_role,
+        )
+
+        pipeline.add_stage(
+            DevStage(
+                self,
+                "dev",
+                env=cdk.Environment(account=constants.DEV_ACCOUNT_ID, region=constants.DEV_REGION),
+            )
+        )
+
+        pipeline.add_stage(
+            PreProdStage(
+                self,
+                "preprod",
```

Comment:
One more observation: deploying in a single account fails with (preprod-endpoint stack):
```
abalone-train-1-p-r6smilm17bgv-endpoint-20240711142847 already exists in stack arn:aws:cloudformation:us-east-1:...:stack/dev-endpoint/14573790-3f92-11ef-b234-0e667f8e3763
```
Looks like we will need to add stage name to the endpoint name to avoid collisions. In addition, we will need to append SM project name to the endpoint stack name.